### DISCUSSION TITLE
Evaluation of the Jacobian in the QuasiStaticNOX solver

### DISCUSSION THREAD

**Discussion by jritta** [peridigm/peridigm#149] [Q&A] [answered]
Hi community,

when I go through the code, I wonder where in the *QuasiStaticNOX* solver the Jacobian is calculated?

For the *Implicit* solver this is done in the function `void PeridigmNS::Peridigm::computeImplicitJacobian(double beta, double dt)` and for the *QuasiStatic* solver it is done by directly call `modelEvaluator->evalJacobian(workset)`.

For the *QuasiStaticNOX* I did not find anything equivalent. Does anyone know where the evaluation is performed? Or is this implicitly done by the `NOX::Solver` class when the method `step()` is called?

I would appreciate any help!

Best,
Jonas

**Comment by johntfoster** [comment, marked as answer]
[Here](https://github.com/peridigm/peridigm/blob/cb9979dd5c83e842b130b147c44714be2dce28b3/src/core/Peridigm.cpp#L1767), which calls [evaluateNOX](https://github.com/peridigm/peridigm/blob/cb9979dd5c83e842b130b147c44714be2dce28b3/src/core/Peridigm.cpp#L1817).  Note that usually when using the NOX solver we don't explicitly construct the Jacobian, but rather use the matrix free option.  This setting can be set using the NOX parameter settings from the input deck.

**Reply by jritta** [reply]
Thank you very much for your reply @johntfoster!
I see your point, but the function [`bool PeridigmNS::Peridigm::computeJacobian`](https://github.com/peridigm/peridigm/blob/cb9979dd5c83e842b130b147c44714be2dce28b3/src/core/Peridigm.cpp#L1767) is not always called by the function [NOXQuasiStatic solver](https://github.com/peridigm/peridigm/blob/cb9979dd5c83e842b130b147c44714be2dce28b3/src/core/Peridigm.cpp#L1817).
I followed the function calls and [`bool PeridigmNS::Peridigm::computeJacobian`](https://github.com/peridigm/peridigm/blob/cb9979dd5c83e842b130b147c44714be2dce28b3/src/core/Peridigm.cpp#L1767) is only called by [`void PeridigmNS::Peridigm::jacobianDiagnostics`](https://github.com/peridigm/peridigm/blob/cb9979dd5c83e842b130b147c44714be2dce28b3/src/core/Peridigm.cpp#L2055) and this one is only called when the parameter [`Jacobian Diagnostics`](https://github.com/peridigm/peridigm/blob/cb9979dd5c83e842b130b147c44714be2dce28b3/src/core/Peridigm.cpp#L2179) is set to `true` in the input deck (by the way, this one blows up the memeory by continously allocate more memeroy. I guess there is somewhere a memory leak)

This is why I wonder how the Jacobian (`tangent`) is calculated for the NOX solver.

Regarding the matrix free option: Independent of using the NOX solver option `NOXQuasiStatic` --> `Direction` --> `Newton` --> `Linear Solver` --> `Jacobian Operator: "Matrix-Free"` the `tangent` is allocated by Peridigm:
This call https://github.com/peridigm/peridigm/blob/cb9979dd5c83e842b130b147c44714be2dce28b3/src/core/Peridigm.cpp#L239 sets the flag `implicitTimeIntegration = true` which is used by https://github.com/peridigm/peridigm/blob/cb9979dd5c83e842b130b147c44714be2dce28b3/src/core/Peridigm.cpp#L260 to set the flag `allocateTangent = true` and this one triggers the [allocation](https://github.com/peridigm/peridigm/blob/cb9979dd5c83e842b130b147c44714be2dce28b3/src/core/Peridigm.cpp#L763).

**Reply by jritta** [reply]
I think you are right @johntfoster. I did not realize that the `Peridigm` object inherits from the `NOX::Epetra::Interface::Required`, `NOX::Epetra::Interface::Jacobian`, and `NOX::Epetra::Interface::Preconditioner` objects and thus the pure virtual method `bool computeJacobian(const Epetra_Vector& x, Epetra_Operator& Jac)` is implemented in the `Peridigm` object which is required by the class `NOX::Epetra::Interface::Jacobian`.
I hope this helps me to find the reason why the NOX solver fails for very large systems (#144). Do you happen to have an idea or a hint where I could look for the possible bug?
